Consistent variable names for rings

orlitzky opened the pull request:
Fix https://github.com/sagemath/sage/issues/34441 by making the `variable_names()` method for `ZZ`, `QQ`, and `AA` return an empty tuple.

orlitzky committed:
src/sage/structure/category_object.pyx: make variable_names() consistent

The variable_names() method is documented to return a tuple of
strings, but if there are no variable names it currently raises an
error. This is inconsistent with the documentation, but maybe more
importantly, inconsistent with the implementation in some rings like
RR where an empty tuple is returned.

This commit changes the "no variables" result to the empty tuple (),
with the justification that it's much easier to loop through or check
for an empty tuple than it is an either-a-tuple-or-an-error.

orlitzky committed:
src/sage/rings: ZZ, QQ, and AA have no variables

Fix issue 34441 by explicitly assigning an empty tuple of names.

orlitzky committed:
src/sage/rings:: ZZ, QQ, and AA have no variables

Fix issue 34441 by removing 'x' from their lists of variables.

orlitzky committed:
src/sage/groups/finitely_presented.py: variable_names() fallout

One test in this file is trying to inject the (nonexistent) variable
names from the finitely-presented group A5, and is expecting an
error. The implementation of variable_names() has been changed,
however, and now returns an empty tuple rather than an error when
there are no variables. As a result, A5.inject_variables() injects no
variables into the session, i.e. does nothing.

It seems like a weird thing to be testing in the first place, so this
commit removes the test rather than updating the expected output.

In another spot, the call to variable_names() is trying to catch the
ValueError that we no longer throw. This has been changed to an "if",
to catch the empty tuple instead.

orlitzky committed:
src/sage/modules/free_module_element.pyx: try/except -> if not

Rather than returning an error, the variable_names() method now
returns an empty tuple when there are no variable names. As a result,
we have to change one try/except clause in this file to "if not".

orlitzky committed:
src/sage/rings/quotient_ring.py: try/except -> if (not)

Rather than returning an error, the variable_names() method now
returns an empty tuple when there are no variable names. As a result
we have to change some try/except clauses in this file to if (not).

orlitzky committed:
src/sage/algebras/lie_algebras/lie_algebra.py: try/except -> if not

Rather than returning an error, the variable_names() method now
returns an empty tuple when there are no variable names. As a result
we have to change one try/except clause in this file to "if not".

orlitzky committed:
src/sage/rings/qqbar.py: drop AA.__new__() method

This method was added to support unpickling of older non-singleton
versions of the AA object, but the test for it happens to be loading
such an old version of AA into scope, making it difficult to doctest
newly-fixed bugs. Since this class has been a singleton since commit
dfbb8bba in 2014, I think it is time to lay this method and its test
to rest.

orlitzky committed:
src/sage/rings/qqbar.py: drop AA.__new__() method

This method was added to support unpickling of older non-singleton
versions of the AA object, but the test for it happens to be loading
such an old version of AA into scope, making it difficult to doctest
newly-fixed bugs. Since this class has been a singleton since commit
dfbb8bba in 2014, I think it is time to lay this method and its test
to rest.

orlitzky committed:
src/sage/rings/quotient_ring_element.py: try/except -> if/else

Rather than returning an error, the variable_names() method now
returns an empty tuple when there are no variable names. As a result
we have to change one try/except clause in this file to if/else.

orlitzky committed:
src/sage/rings/asymptotic/misc.py: try/except -> if

Rather than returning an error, the variable_names() method now
returns an empty tuple when there are no variable names. As a result
we have to change one try/except clause in this file to an "if".

Note: there were two ValueErrors being caught by this one "except"
clause, one from variable_names() and one from abbreviate(). Only
the first becomes an "if" statement.

orlitzky committed:
src/sage/algebras/lie_algebras/quotient.py: try/except -> if

Rather than returning an error, the variable_names() method now
returns an empty tuple when there are no variable names. As a result
we have to change one try/except clause in this file to an "if".

orlitzky committed:
src/sage/.../lie_conformal_algebra_element.py: drop one try/except

Rather than returning an error, the variable_names() method now
returns an empty tuple when there are no variable names. As a result
we can drop one try/except clause in this file completely.

orlitzky commented:
Hi everyone who was CC'ed on https://github.com/sagemath/sage/issues/34441. I've fixed this inconsistency by returning an empty tuple rather than an error because it's usually much cleaner to loop through an empty tuple (i.e. do nothing) or check for one with `if foo` than it is to try/except a `ValueError`.

The fix itself was relatively easy. Afterwards there were a few doctest failures that needed tweaks, and then a somewhat larger number of apparently untested try/except clauses around `variable_names()` that I've converted to if/else.

mantepse commented on src/sage/rings/qqbar.py:
```diff
@@ -1075,40 +1075,6 @@ class AlgebraicRealField(Singleton, AlgebraicField_common, sage.rings.abc.Algebr
-
-        TESTS::
-
-            sage: s = loads(b'x\x9cmQ\xcbR\x141\x14\xad\x11A\x083\xe2\x03T|'
-            ....: b'\x82l`\xd3\xff\xe0\x86\x8de/\xba*\xcb\xa9[\xe9\xf4'
-            ....: b'\xa5;e:=\'I+,\xa6J\x17B\xf9\xd7f\x08\xe2s\x95\xa4\xee9\xf7<'
-            ....: b'\xf2\xe5\x8e\x0e\xaa\xe5"D?\xea8z.\x9a\x0b\xa7z\xa3I[\x15'
-            ....: b'\x82\xf8\xf3\x85\xc9\xb1<xg[\xae\xbd2\xbabeO\r\xdb\x86>\x9b'
-            ....: b'\xd8\x91V\x91\xdb\xc1_\xe0f\xa57\xae\r\x05P+/\xfe\xe5\x08'
-            ....: b'\xaci\xa2z46\x1aG$Z\x8e*F/p\xf7oC\xa33\x18\x99</<\x07v\tf'
-            ....: b'\x06\'F\xe7\xb9\x195\x0b\xacg\xc2\x8d\xbc\xe1P\x9c\xad\x04'
-            ....: b'\x828\xcd\x076N\x96W\xb8WaSN\x17\xca\xa7\r9\r\xb6.+\x88Kl'
-            ....: b'\x97e\xb7\x16+LO\xbeb\xb6\xc4\xfdc)\x88\xfb\x9a\x9b&\x05'
-            ....: b'\xc0N)wI\x0f\xee\x13\xfbH=\xc7nh(U\xc2xP\xca\r\xd2\x8d'
-            ....: b'\x8a\n\x0fK\xb9\xf5+\xfe\xa3n3MV\x98\x80\xc7rr\xfe\r\xbbr'
-            ....: b'\x9bZv\xecU\x1c|\xc0\xde\x12O\xe4:\xd5*0\x9ev3\xb9C\x0b'
-            ....: b'\xa3?Z\xa6\xa4\x11R6<{?I\xa2l\xb9\xbf6;\xb8\\\xc6\xe0\xb1'
-            ....: b'\x9f\xb3\xf6&\xe8\xe2,\xb3R\x13\xf9\xf2\xe1\xda\x9c\xc0s'
-            ....: b'\xb9\xf7?.\xe1E7\xeb\xa6W\x15^&\x80q&\x1aeo\x93Y\x13"^\xcd'
-            ....: b'\xf1Z\xee\xdf\x92W\x18Z\xa4\xa6(\xd7\x867\xdf\x93\xad\x9fL'
-            ....: b'\xa5W\xff\x90\x89\x07s\x1c\xfe6\xd2\x03{\xcdy\xf4v\x8e\xa3'
-            ....: b'\xb1.~\x000\xc2\xe0\xa1')
-            sage: s is AA
-            True
```
I don't think you can remove this test.

orlitzky commented:
> I don't really agree with point 2 of the PR. There is a difference between no variable names and not having assigned any. I think changing the default to be `()` is good, but this is different than saying we know there are no variables.

I understand the distinction, but I'm not sure that it matters. This hits on point 2. As far as the docs are concerned, the `variable_names()` method is supposed to...

```python
def variable_names(self):
    """                                                                                                                                                    
    Return the list of variable names corresponding to the generators.                                                                                     
                                                                                                                                                               
    OUTPUT: a tuple of strings 
    ...
    """
```

It doesn't mention that a `ValueError` may be thrown, so in that sense, returning the empty tuple is more consistent to me.

Ignoring for the moment that `names=()` does not work without further code changes (see the OP), we _could_ keep the `ValueError`, and require subclasses to provide the empty tuple explicitly. But, the other half of what I was getting at in (2) is that, from what I saw, no subclasses care about the distinction. They're happily leaving `._names` unassigned, and then catching the `ValueError` to indicate that there are no variables. One typical example:

```python
try:
    names = p.latex_variable_names()
except ValueError:
    names = None
if names:
    # do stuff
```

After changing the `ValueError` to `()`, this becomes simply:

```python
names = p.latex_variable_names()
if names:
    # do stuff
```

This is much more natural to me in the absence of any need to detect `._names` being unset entirely.

> In particular, we should allow for delayed assignment of variables because you might need to fully call `super().__init__` on an instance (which would call `Parent.__init__()`) before setting the variable names.

This has not changed as far as I know. The `variable_names()` method will return any empty tuple if `._names` is unset, but `._names` is never assigned until you call `_assign_names()`. If you don't pass any names to `__init__()`, you should still be able to call `_assign_names()` later on.

> We obviously should not allow a user to have a hook to override the variable names (which would imply more mutability than I think we should allow). Will this also work if `Parent.__init__()` is called twice?

The behavior is unchanged, insofar that:

1. Calling `Parent.__init__()` twice with the same `names=` is fine, as is passing no names the first time and whatever names you want the second.
2. Calling `Parent.__init__()` twice with different `names=` will eventually cause an exception in `_assign_names`, since it does not allow you to change existing names.

tscrim commented:
Sorry for loosing track of this. I think I misunderstood the intent of this PR a bit when reading it before. So I'm also apologizing for any confusion.

My point is that you cannot distinguish the cases when `()` indicates when variables are not set and when there are no variables. You need to be able to distinguish between these two cases. Yes, nearly all cases will be corner cases, but it is a mathematically important distinction. Compare: some polynomial ring implementation that has not set its variables versus a polynomial ring with no variables (i.e., just the base ring). So if it is unset, rather than a `ValueError` we could change the behavior of `variable_names()` to return `None`. Although I think raising an error is still a legitimate way to handle that case...

In any case, this is a backwards incompatible change, and hopefully it won't affect too many people for too much code. Perhaps we can issue a deprecation warning when no variables are set but one calls `variable_names()` without too much disruption?

mantepse commented:
I don't quite understand yet: if an object has variables, they are not yet set, and accessed, then that should be an error, and it shouldn't be caught (except in for very special reason, such as setting the variables), should it?

mantepse commented:
I would really prefer to be precise here, because as we can see from the `ParentWithGens` disaster, convenience is bound to lead to problems later.

orlitzky commented:
> Sorry for loosing track of this. I think I misunderstood the intent of this PR a bit when reading it before. So I'm also apologizing for any confusion.
> 
> My point is that you cannot distinguish the cases when `()` indicates when variables are not set and when there are no variables. You need to be able to distinguish between these two cases. Yes, nearly all cases will be corner cases, but it is a mathematically important distinction.

You can check `._names` directly if you really need to know... what scenario do you have in mind?

If an object with names forgets to assign them, then `variable_names()` will return a wrong answer (the empty tuple) instead of raising an error. That is harder to notice, but still an easy-to-test bug in the constructor.

> Compare: some polynomial ring implementation that has not set its variables versus a polynomial ring with no variables (i.e., just the base ring).

In the base ring, `variable_names()` will immediately be correct without having to call `_assign_names()`. (The need to assign names in a ring without any names may be how we wound up with an 'x' in the ring of integers.) In a polynomial ring with variables, `variable_names()` will simply be wrong until the names are assigned.

Note that all of the "fallout" updates made to the sage library are cases where we _already_ wanted to check for "no variables," but were checking for "variables not assigned" instead. I don't remember what goes wrong, but in the original PR description it looks like I ran into trouble specifying `names=()`, meaning that it was already difficult to distinguish the two cases.


> In any case, this is a backwards incompatible change, and hopefully it won't affect too many people for too much code. Perhaps we can issue a deprecation warning when no variables are set but one calls `variable_names()` without too much disruption?

Though this PR itself refutes the ideal, anyone calling a method that returns a tuple should be prepared to handle an empty tuple :)

It is user-facing, but agrees with the documentation and doesn't change any types. I wouldn't be opposed to a warning though.

orlitzky commented:
> I don't quite understand yet: if an object has variables, they are not yet set, and accessed, then that should be an error, and it shouldn't be caught (except in for very special reason, such as setting the variables), should it?

The variable names aren't being accessed directly, they are being returned from a method that promises to return a tuple and says nothing about ValueErrors. And for "historical reasons" (meaning: I don't know) in many cases "no variables" was specified by leaving them unset. You can see in the diff many places where the error was caught and handled in a context where it means "no variables."

mantepse commented:
> > I don't quite understand yet: if an object has variables, they are not yet set, and accessed, then that should be an error, and it shouldn't be caught (except in for very special reason, such as setting the variables), should it?
> 
> The variable names aren't being accessed directly, they are being returned from a method that promises to return a tuple and says nothing about ValueErrors. And for "historical reasons" (meaning: I don't know) in many cases "no variables" was specified by leaving them unset. You can see in the diff many places where the error was caught and handled in a context where it means "no variables."

Yes.  In my opinion, the bug is in catching the error.  This seems completely wrong to me.  `variable_names` should throw an error, if `self._names is None`.

I also completely agree that `ZZ.variable_names()` should return the empty tuple.

orlitzky commented:
> Yes. In my opinion, the bug is in catching the error. This seems completely wrong to me. `variable_names` should throw an error, if `self._names is None`.

This is where it gets tricky... the documentation says that `variable_names()` will always return a tuple. And apparently (per my original description, it's gone from memory), there are several cases where passing `names=()` does not work. In those cases, the names are left unspecified to indicate "no names." We then end up with consumers doing try/catch on the ValueError to see if there are no names -- the diff fixes several of these after getting rid of the ValueError.

There is probably a better design that would clean this all up, but

> I've left it alone rather than create a diff that touches 100 files.

mantepse commented:
> > Yes. In my opinion, the bug is in catching the error. This seems completely wrong to me. `variable_names` should throw an error, if `self._names is None`.
> 
> This is where it gets tricky... the documentation says that `variable_names()` will always return a tuple. And apparently (per my original description, it's gone from memory), there are several cases where passing `names=()` does not work. In those cases, the names are left unspecified to indicate "no names." We then end up with consumers doing try/catch on the ValueError to see if there are no names -- the diff fixes several of these after getting rid of the ValueError.

Can you say what fails if we keep the `ValueError`?

tscrim commented:
I would say very few methods indicate when they raise am error, even during "normal" operations. In particular, I  would not treat that methods documentation as gospel.

mantepse commented:
Is
```
sage: QQbar.variable_names()
('I',)
```
really what we want?

mantepse commented:
With the patch below applied, I don't see any obvious problems.  I cannot test all of the library, because my conda install has problems with `sage.features.FeatureNotPresentError: database_cremona_mini_ellcurve is not available.`
```diff
diff --git a/src/sage/rings/integer_ring.pyx b/src/sage/rings/integer_ring.pyx
index 28cbdc8988e..b544c6f346c 100644
--- a/src/sage/rings/integer_ring.pyx
+++ b/src/sage/rings/integer_ring.pyx
@@ -299,7 +299,7 @@ cdef class IntegerRing_class(CommutativeRing):
         """
         cat = (EuclideanDomains(), DedekindDomains(),
                InfiniteEnumeratedSets().Metric(), NoetherianRings())
-        Parent.__init__(self, base=self, normalize=False,
+        Parent.__init__(self, base=self, normalize=False, names=(),
                         category=cat)
         self._populate_coercion_lists_(init_no_parent=True,
                                        convert_method_name='_integer_')
diff --git a/src/sage/rings/qqbar.py b/src/sage/rings/qqbar.py
index cb7f393d768..55f1351bb05 100644
--- a/src/sage/rings/qqbar.py
+++ b/src/sage/rings/qqbar.py
@@ -1086,7 +1086,8 @@ class AlgebraicRealField(Singleton, AlgebraicField_common, sage.rings.abc.Algebr
 
         """
         from sage.categories.fields import Fields
-        AlgebraicField_common.__init__(self, self, normalize=False, category=Fields().Infinite())
+        AlgebraicField_common.__init__(self, self, normalize=False, names=(),
+                                       category=Fields().Infinite())
         self._populate_coercion_lists_([ZZ, QQ])
 
     def _element_constructor_(self, x):
diff --git a/src/sage/rings/rational_field.py b/src/sage/rings/rational_field.py
index 8335cbc2b82..c8aa748a9ea 100644
--- a/src/sage/rings/rational_field.py
+++ b/src/sage/rings/rational_field.py
@@ -231,7 +231,7 @@ class RationalField(Singleton, number_field_base.NumberField):
         """
         from sage.categories.basic import QuotientFields
         from sage.categories.number_fields import NumberFields
-        Parent.__init__(self, base=self,
+        Parent.__init__(self, base=self, normalize=False, names=(),
                         category=[QuotientFields().Metric(),
                                   NumberFields()])
         self._populate_coercion_lists_(init_no_parent=True)
diff --git a/src/sage/structure/category_object.pyx b/src/sage/structure/category_object.pyx
index 918d60c8884..04958aaba87 100644
--- a/src/sage/structure/category_object.pyx
+++ b/src/sage/structure/category_object.pyx
@@ -461,11 +461,9 @@ cdef class CategoryObject(SageObject):
 
     def variable_names(self):
         """
-        Return the variable names corresponding to our generators.
+        Return the list of variable names corresponding to the generators.
 
-        OUTPUT:
-
-        A (possibly empty) tuple of strings.
+        OUTPUT: a tuple of strings
 
         EXAMPLES::
 
@@ -481,16 +479,10 @@ cdef class CategoryObject(SageObject):
             sage: T.<x> = InfinitePolynomialRing(ZZ)
             sage: T.variable_names()
             ('x',)
-
-        ::
-
-            sage: NN.variable_names()
-            ()
-
         """
         if self._names is not None:
             return self._names
-        return ()
+        raise ValueError("variable names have not yet been set using self._assign_names(...)")
 
     def variable_name(self):
         """
@@ -532,11 +524,13 @@ cdef class CategoryObject(SageObject):
         # We cannot assume that self *has* _latex_variable_names.
         # But there is a method that returns them and sets
         # the attribute at the same time, if needed.
-
-        old = None, None
-        if self.variable_names():
+        # Simon King: It is not necessarily the case that variable
+        # names are assigned. In that case, self._names is None,
+        # and self.variable_names() raises a ValueError
+        try:
             old = self.variable_names(), self.latex_variable_names()
-
+        except ValueError:
+            old = None, None
         self._names, self._latex_names = names, latex_names
         return old
```

mantepse commented:
Here are the problems I see:
* what should `SR.variable_names()` do?  Currently, it throws `ValueError: variable names have not yet been set using self._assign_names(...)`.
* what should `InfinitePolynomialRing(QQ, "y").variable_names()` do?  Currently, it returns `('y',)`.

orlitzky commented:
> Can you say what fails if we keep the `ValueError`?

I don't think anything fails, but then we retain lots of code that needs to catch `ValueError("variable names have not yet been set using self._assign_names(...)")` if it wants to check for "no variables." And in some cases (if my description is to be believed), there is no good way to distinguish between the two.

I may have bitten off too much here. In most cases it should be possible to fix the variables-are-wrong bugs without refactoring. Everyone can agree on that much.

orlitzky commented:
> Is
> 
> ```
> sage: QQbar.variable_names()
> ('I',)
> ```
> 
> really what we want?

Sooner or later I'm going to have to admit that I don't know what a variable is.

mantepse commented:
> > Is
> > ```
> > sage: QQbar.variable_names()
> > ('I',)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > really what we want?
> 
> Sooner or later I'm going to have to admit that I don't know what a variable is.

I feel you.  I think that sage is way too sloppy here.  It's simply not a good idea that every parent should have generators, and even less so, a finite set of generators.  Same goes for `base_ring`.

orlitzky commented:
I've replaced this with a minimal fix.

orlitzky commented:
I think the CI is OK now. I also tried to fix

```
sage: GF(2,1).variable_names()
('x',)
sage: Zmod(4).variable_names()
('x',)
```

but it caused a cascade of new problems.

mantepse commented:
The change in `ParentWithGens` will (obviously) conflict with #42232 and #42245.

orlitzky commented:
> The change in `ParentWithGens` will (obviously) conflict with #42232 and #42245.

I'll put this on hold for a few weeks then, thanks for the heads up.

mantepse commented:
Could you resurrect this now?

orlitzky committed:
src/sage/rings: assign ZZ, QQ, and AA names via constructor

Rather than a subsequent call to assign_names(), it seems that now
we are able to pass an empty tuple to the Parent __init__ method.

orlitzky commented:
Attempting to fix `GF(p)` and `Zmod(n)` now causes noticeably less chaos, but I can work on those in a follow-up.

mantepse commented:
> Attempting to fix `GF(p)` and `Zmod(n)` now causes noticeably less chaos, but I can work on those in a follow-up.

As you prefer!

orlitzky committed:
src/sage/rings/rational_field.py: don't normalize names=() in QQ

This should be less work, and more efficient.

mantepse reviewed (approved):
thank you!

orlitzky commented:
no problem, thanks for your persistence
